File: dashboard/tabs/recommendations.py
"""Recommendations tab: on-demand price recommendations for a property
across a daily/weekly/monthly date range, reusing the exact same
explainability payload as Scenario Simulator/Rate Calendar/Approval Queue
(confidence breakdown, arm probabilities, guardrail-excluded arms) for
every day in the range - then Approve, Modify (override), or Reject each
day's recommendation directly from this view. Optionally override context
(occupancy, pace, comp set, events, etc. - same fields as Scenario
Simulator's what-if panel) applied identically to every day, to see how
the recommendation reacts to a specific demand scenario before publishing.

Generating the range is always side-effect-free (POST /recommendations,
persist=false - same as Scenario Simulator's preview). Only clicking
Approve/Modify & Approve actually publishes a real decision (POST /score),
exactly like Scenario Simulator's "Publish as live decision" button, just
applied per-day across the whole range. Reject is a local, UI-only dismissal
- there is nothing to reject server-side for a recommendation that was
never published."""
from __future__ import annotations

import datetime as dt

import pandas as pd
import streamlit as st

from dashboard.api_client import get_properties, get_property_config, get_recommendations, override_decision, score
from dashboard.context_panel import render_context_panel

RANGE_PRESETS = {"Daily (1 day)": 1, "Weekly (7 days)": 7, "Monthly (30 days)": 30}
SEGMENT_OPTIONS = ["transient", "corporate", "group", "leisure"]


def render():
    st.header("🗓️ On-Demand Recommendations")
    st.caption(
        "Generate price recommendations for a property across a date range - daily, weekly, or "
        "monthly - each with the full explainability behind it. Generating is side-effect-free; "
        "use Approve / Modify & Approve / Reject per day to act on individual recommendations."
    )

    properties = get_properties()
    prop_options = {p["property_id"]: f"{p['name']} ({p['property_id']})" for p in properties}

    col1, col2, col3 = st.columns(3)
    with col1:
        property_id = st.selectbox(
            "Property", options=list(prop_options), format_func=lambda x: prop_options[x], key="rec_property"
        )

    config = get_property_config(property_id)
    room_type_options = config.get("room_types") or ["standard"]
    rate_plan_options = config.get("rate_plans") or ["bar_flexible"]
    with col2:
        room_type = st.selectbox("Room type", options=room_type_options, key="rec_room_type")
    with col3:
        rate_plan = st.selectbox("Rate plan", options=rate_plan_options, key="rec_rate_plan")

    col4, col5, col6 = st.columns(3)
    with col4:
        preset = st.radio("Range", options=list(RANGE_PRESETS) + ["Custom"], key="rec_preset")
    with col5:
        start_date = st.date_input("Start date", value=dt.date.today() + dt.timedelta(days=1), key="rec_start")
    with col6:
        if preset == "Custom":
            end_date = st.date_input("End date", value=start_date + dt.timedelta(days=6), key="rec_end")
        else:
            end_date = start_date + dt.timedelta(days=RANGE_PRESETS[preset] - 1)
            st.text_input("End date", value=str(end_date), disabled=True)

    los_nights = st.number_input("Length of stay (nights)", min_value=1, max_value=14, value=2, key="rec_los")

    st.divider()
    override_enabled = st.checkbox(
        "🔧 Override context for this range (apply the same manual context to every day generated below)",
        key="rec_override_enabled",
    )
    context_overrides = None
    if override_enabled:
        st.caption(
            "Applied identically to every day below, overriding the naturally simulated context - see "
            "how the recommendation reacts to a specific demand scenario before publishing/approving "
            "anything. Leave unchecked to use the naturally simulated context per day (default)."
        )
        event_flag = st.checkbox("Local Event Active", value=False, key="rec_override_event_flag")
        oc1, oc2, oc3 = st.columns(3)
        with oc1:
            occupancy_pct = st.slider("Occupancy %", 0.0, 100.0, 55.0, key="rec_override_occupancy")
            adr_trend_pct = st.slider("ADR Trend %", -20.0, 20.0, 0.0, key="rec_override_adr_trend")
            remaining_inventory_pct = st.slider("Remaining Inventory %", 0.0, 100.0, 40.0, key="rec_override_inventory")
        with oc2:
            pace_vs_stly_pct = st.slider("Pace vs Last Year %", -30.0, 30.0, 0.0, key="rec_override_pace")
            pickup_last_7d = st.slider("Pickup (7d, rooms)", 0.0, 50.0, 5.0, key="rec_override_pickup")
            segment = st.selectbox("Guest Segment", options=SEGMENT_OPTIONS, key="rec_override_segment")
        with oc3:
            event_intensity = st.slider(
                "Event Intensity", 0.0, 1.0, 0.0, disabled=not event_flag, key="rec_override_event_intensity"
            )
            comp_set_avg_rate = st.number_input(
                "Comp Set Avg Rate ($)", min_value=0.0, value=200.0, key="rec_override_comp_rate"
            )

        context_overrides = {
            "occupancy_pct": occupancy_pct,
            "adr_trend_pct": adr_trend_pct,
            "remaining_inventory_pct": remaining_inventory_pct,
            "pace_vs_stly_pct": pace_vs_stly_pct,
            "pickup_last_7d": pickup_last_7d,
            "segment": segment,
            "event_flag": event_flag,
            "event_intensity": event_intensity if event_flag else 0.0,
            "comp_set_avg_rate": comp_set_avg_rate,
        }

    request_key = (
        property_id, room_type, rate_plan, str(start_date), str(end_date), los_nights,
        tuple(sorted(context_overrides.items())) if context_overrides else None,
    )

    if st.button("Generate recommendations", type="primary"):
        if end_date < start_date:
            st.error("End date must be on or after start date.")
        else:
            with st.spinner(f"Scoring {(end_date - start_date).days + 1} day(s)..."):
                results = get_recommendations(
                    property_id, room_type, rate_plan, str(start_date), str(end_date), los_nights,
                    context_overrides=context_overrides,
                )
            st.session_state["rec_last_key"] = request_key
            st.session_state["rec_last_results"] = results
            st.session_state["rec_last_overrides"] = context_overrides
            st.session_state["rec_actions"] = {}

    if st.session_state.get("rec_last_key") != request_key:
        st.info("Pick a property/room/rate/date range above and click **Generate recommendations**.")
        return

    results = st.session_state.get("rec_last_results") or []
    if not results:
        return

    actions = st.session_state.setdefault("rec_actions", {})

    if st.session_state.get("rec_last_overrides"):
        st.info("🔧 Showing recommendations generated WITH manual context overrides applied (see checkbox above).")

    st.subheader(f"Summary - {len(results)} day(s)")
    summary_df = pd.DataFrame(
        [
            {
                "stay_date": r["stay_date"],
                "arm": r["chosen_arm_label"],
                "offset": f"{r['chosen_arm_offset_pct'] * 100:+.1f}%",
                "price": r["published_price"],
                "confidence": r["confidence_label"],
                "requires_approval": r["requires_approval"],
                "status": actions.get(r["stay_date"], {}).get("status", "pending review"),
            }
            for r in results
        ]
    )
    st.dataframe(summary_df, use_container_width=True)

    approver = st.text_input("Approver name", value="revenue_manager", key="rec_approver")

    st.subheader("Day-by-day detail")
    for r in results:
        day = r["stay_date"]
        done = actions.get(day)
        label = (
            f"{day} | {r['chosen_arm_label']} ({r['chosen_arm_offset_pct'] * 100:+.1f}%) -> "
            f"${r['published_price']:.2f} [{r['confidence_label']}]"
        )
        if done:
            label += f"  —  {done['status']}"
        with st.expander(label):
            render_context_panel(r.get("context", {}), expanded=False)
            with st.expander("All arm probabilities"):
                st.bar_chart({a["label"]: a["probability"] for a in r["all_arms"]})
            if r["excluded_arms"]:
                with st.expander("Guardrail-excluded arms"):
                    st.json(r["excluded_arms"])

            if done:
                st.success(f"{done['status']} - decision_id `{done.get('decision_id', 'n/a')}`")
                continue

            c1, c2, c3 = st.columns(3)
            with c1:
                if st.button("✅ Approve", key=f"rec_approve_{day}"):
                    live = score(property_id, room_type, rate_plan, day, dry_run=False, los_nights=los_nights)
                    status = "Published (pending approval)" if live.get("requires_approval") else "Published (auto-published)"
                    actions[day] = {"status": status, "decision_id": live.get("decision_id")}
                    st.rerun()
            with c2:
                override_price = st.number_input(
                    "Modify price ($)", min_value=0.0, value=float(r["published_price"]), key=f"rec_price_{day}"
                )
                if st.button("✏️ Modify & Approve", key=f"rec_modify_{day}"):
                    live = score(property_id, room_type, rate_plan, day, dry_run=False, los_nights=los_nights)
                    override_decision(live["decision_id"], approver, override_price)
                    actions[day] = {"status": "Published (modified & approved)", "decision_id": live.get("decision_id")}
                    st.rerun()
            with c3:
                if st.button("🚫 Reject", key=f"rec_reject_{day}"):
                    actions[day] = {"status": "Rejected (not published)", "decision_id": None}
                    st.rerun()

File: frontend/js/pages/recommendations.js
/**
 * Recommendations page - on-demand daily/weekly/monthly price recommendations.
 * Full context overrides (all API-supported fields) + context detail popover per day.
 */
const RecommendationsPage = {
    results: [],
    chart: null,

    async render(container) {
        container.innerHTML = this.getHTML();
        this.bindEvents();
        if (App.selectedProperty) {
            document.getElementById('rec-property').value = App.selectedProperty;
            this.loadPropertyConfig(App.selectedProperty);
        }
    },

    getHTML() {
        return `
        <div class="page-content">
            <div class="d-flex justify-content-between align-items-center mb-3">
                <div>
                    <h4 class="mb-0"><i class="bi bi-calendar-range me-2"></i>On-Demand Recommendations</h4>
                    <small class="text-muted">Generate pricing recommendations, review context per day, then approve or adjust.</small>
                </div>
            </div>
            <div class="row">
                <div class="col-md-4">${this.getControlsHTML()}</div>
                <div class="col-md-8"><div id="rec-results">${this.getEmptyState()}</div></div>
            </div>
            ${this.getContextModalHTML()}
        </div>`;
    },

    getControlsHTML() {
        return `
            <div class="card shadow-sm">
                <div class="card-header bg-light"><strong>Generate Recommendations</strong></div>
                <div class="card-body">
                    <div class="mb-2">
                        <label class="form-label form-label-sm">Property</label>
                        <select class="form-select form-select-sm" id="rec-property">
                            ${App.properties.map(p => `<option value="${p.property_id}">${p.name}</option>`).join('')}
                        </select>
                    </div>
                    <div class="row mb-2">
                        <div class="col-6">
                            <label class="form-label form-label-sm">Room Type</label>
                            <select class="form-select form-select-sm" id="rec-room-type">
                                <option value="standard">Standard</option>
                                <option value="deluxe">Deluxe</option>
                                <option value="suite">Suite</option>
                            </select>
                        </div>
                        <div class="col-6">
                            <label class="form-label form-label-sm">Rate Plan</label>
                            <select class="form-select form-select-sm" id="rec-rate-plan">
                                <option value="bar_best_available">BAR (Best Available)</option>
                                <option value="government_military">Government/Military</option>
                                <option value="senior">Senior</option>
                                <option value="special_offer">Special Offer</option>
                            </select>
                        </div>
                    </div>
                    <div class="row mb-2">
                        <div class="col-6">
                            <label class="form-label form-label-sm">LOS (nights)</label>
                            <input type="number" class="form-control form-control-sm" id="rec-los" value="2" min="1" max="14">
                        </div>
                    </div>
                    <hr>
                    <h6 class="text-muted small">Date Range</h6>
                    <div class="btn-group w-100 mb-2" role="group">
                        <button class="btn btn-sm btn-outline-primary active rec-range-btn" data-days="7">7d</button>
                        <button class="btn btn-sm btn-outline-primary rec-range-btn" data-days="14">14d</button>
                        <button class="btn btn-sm btn-outline-primary rec-range-btn" data-days="30">30d</button>
                    </div>
                    <div class="row mb-2">
                        <div class="col-6">
                            <input type="date" class="form-control form-control-sm" id="rec-start" value="${Utils.dateOffset(1)}">
                        </div>
                        <div class="col-6">
                            <input type="date" class="form-control form-control-sm" id="rec-end" value="${Utils.dateOffset(7)}">
                        </div>
                    </div>

                    <hr>
                    <div class="d-flex justify-content-between align-items-center mb-2">
                        <h6 class="text-muted small mb-0">Context Overrides</h6>
                        <button class="btn btn-xs btn-link p-0 text-decoration-none" type="button" data-bs-toggle="collapse" data-bs-target="#rec-overrides-panel">
                            <i class="bi bi-chevron-down"></i> expand
                        </button>
                    </div>
                    <div class="collapse show" id="rec-overrides-panel">
                        ${this.getOverridesHTML()}
                    </div>

                    <button class="btn btn-primary w-100 mt-3" id="rec-generate-btn">
                        <i class="bi bi-lightning-fill"></i> Generate Recommendations
                    </button>
                </div>
            </div>`;
    },

    getOverridesHTML() {
        return `
            <small class="text-muted d-block mb-2">Leave blank = auto-generated from property data. Fill in to override.</small>
            <div class="row g-2">
                <div class="col-6">
                    <label class="form-label form-label-sm">Occupancy %</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-occupancy" placeholder="Auto" min="5" max="99">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">ADR Trend %</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-adr-trend" placeholder="Auto" step="0.1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Pace vs STLY %</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-pace" placeholder="Auto" step="0.1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Pickup (7d)</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-pickup" placeholder="Auto" min="0" step="1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Remaining Inv. %</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-remaining" placeholder="Auto" min="1" max="100">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Comp Set Avg Rate</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-compset-rate" placeholder="Auto" min="50" step="1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Our vs Compset Index</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-compset-index" placeholder="Auto" min="0.5" max="2" step="0.01">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Compset Trend %</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-compset-trend" placeholder="Auto" step="0.1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Compset Dispersion</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-compset-disp" placeholder="Auto" min="0" max="1" step="0.01">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Event Intensity</label>
                    <input type="number" class="form-control form-control-sm" id="rec-ov-event" placeholder="Auto" min="0" max="1" step="0.1">
                </div>
                <div class="col-6">
                    <label class="form-label form-label-sm">Segment</label>
                    <select class="form-select form-select-sm" id="rec-ov-segment">
                        <option value="">Auto</option>
                        <option value="transient">Transient</option>
                        <option value="corporate">Corporate</option>
                        <option value="leisure">Leisure</option>
                        <option value="group">Group</option>
                    </select>
                </div>
            </div>`;
    },

    getContextModalHTML() {
        return `
            <div class="modal fade" id="rec-context-modal" tabindex="-1">
                <div class="modal-dialog modal-lg">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title"><i class="bi bi-info-circle me-2"></i>Full Decision Context</h5>
                            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body" id="rec-context-modal-body"></div>
                    </div>
                </div>
            </div>`;
    },

    getEmptyState() {
        return `<div class="text-center text-muted py-5">
            <i class="bi bi-calendar-plus" style="font-size:3rem"></i>
            <p class="mt-2">Configure parameters and click "Generate Recommendations"<br>
            to preview what the model would price for each day.</p>
            <p class="text-muted small">Preview only — nothing is published until you approve.</p>
        </div>`;
    },

    bindEvents() {
        document.querySelectorAll('.rec-range-btn').forEach(btn => {
            btn.addEventListener('click', () => {
                document.querySelectorAll('.rec-range-btn').forEach(b => b.classList.remove('active'));
                btn.classList.add('active');
                document.getElementById('rec-end').value = Utils.dateOffset(parseInt(btn.dataset.days));
            });
        });
        document.getElementById('rec-property').addEventListener('change', (e) => this.loadPropertyConfig(e.target.value));
        document.getElementById('rec-generate-btn').addEventListener('click', () => this.generate());
    },

    async loadPropertyConfig(propertyId) {
        try {
            const config = await API.getPropertyConfig(propertyId);
            document.getElementById('rec-room-type').innerHTML = config.room_types.map(rt => `<option value="${rt}">${rt}</option>`).join('');
            document.getElementById('rec-rate-plan').innerHTML = config.rate_plans.map(rp => `<option value="${rp}">${rp}</option>`).join('');
        } catch (e) { /* use defaults */ }
    },

    buildOverrides() {
        const ov = {};
        const fields = [
            { id: 'rec-ov-occupancy', key: 'occupancy_pct', type: 'float' },
            { id: 'rec-ov-adr-trend', key: 'adr_trend_pct', type: 'float' },
            { id: 'rec-ov-pace', key: 'pace_vs_stly_pct', type: 'float' },
            { id: 'rec-ov-pickup', key: 'pickup_last_7d', type: 'float' },
            { id: 'rec-ov-remaining', key: 'remaining_inventory_pct', type: 'float' },
            { id: 'rec-ov-compset-rate', key: 'comp_set_avg_rate', type: 'float' },
            { id: 'rec-ov-compset-index', key: 'our_rate_vs_compset_index', type: 'float' },
            { id: 'rec-ov-compset-trend', key: 'compset_rate_trend_pct', type: 'float' },
            { id: 'rec-ov-compset-disp', key: 'compset_dispersion', type: 'float' },
            { id: 'rec-ov-event', key: 'event_intensity', type: 'float' },
        ];
        for (const f of fields) {
            const val = document.getElementById(f.id)?.value;
            if (val !== '' && val !== null && val !== undefined) {
                ov[f.key] = parseFloat(val);
            }
        }
        // event_flag derived from event_intensity
        if (ov.event_intensity !== undefined) {
            ov.event_flag = ov.event_intensity > 0;
        }
        // segment
        const seg = document.getElementById('rec-ov-segment')?.value;
        if (seg) ov.segment = seg;

        return Object.keys(ov).length > 0 ? ov : null;
    },

    async generate() {
        const btn = document.getElementById('rec-generate-btn');
        btn.disabled = true;
        btn.innerHTML = '<span class="spinner-border spinner-border-sm"></span> Generating...';
        try {
            const payload = {
                property_id: document.getElementById('rec-property').value,
                room_type: document.getElementById('rec-room-type').value,
                rate_plan: document.getElementById('rec-rate-plan').value,
                start_date: document.getElementById('rec-start').value,
                end_date: document.getElementById('rec-end').value,
                los_nights: parseInt(document.getElementById('rec-los').value),
                persist: false,
                context_overrides: this.buildOverrides(),
            };
            this.results = await API.getRecommendations(payload);
            this.renderResults();
        } catch (e) {
            Utils.showToast('Failed to generate: ' + e.message, 'error');
        } finally {
            btn.disabled = false;
            btn.innerHTML = '<i class="bi bi-lightning-fill"></i> Generate Recommendations';
        }
    },

    renderResults() {
        if (!this.results.length) {
            document.getElementById('rec-results').innerHTML = '<div class="alert alert-info">No results.</div>';
            return;
        }
        const el = document.getElementById('rec-results');
        el.innerHTML = `
            <div class="row mb-3">
                <div class="col-md-3"><div class="card stat-card"><div class="card-body text-center">
                    <div class="stat-value">${this.results.length}</div><div class="stat-label">Days</div>
                </div></div></div>
                <div class="col-md-3"><div class="card stat-card"><div class="card-body text-center">
                    <div class="stat-value">${Utils.currency(this.avg('published_price'))}</div><div class="stat-label">Avg Price</div>
                </div></div></div>
                <div class="col-md-3"><div class="card stat-card"><div class="card-body text-center">
                    <div class="stat-value">${this.results.filter(r => r.requires_approval).length}</div><div class="stat-label">Need Approval</div>
                </div></div></div>
                <div class="col-md-3"><div class="card stat-card"><div class="card-body text-center">
                    <div class="stat-value">${(this.avg('confidence_score') * 100).toFixed(0)}%</div><div class="stat-label">Avg Confidence</div>
                </div></div></div>
            </div>
            <div class="card shadow-sm mb-3">
                <div class="card-header bg-light d-flex justify-content-between align-items-center">
                    <strong>Price Curve (recommended vs reference)</strong>
                    <button class="btn btn-sm btn-success" id="rec-approve-all"><i class="bi bi-check-all"></i> Approve All</button>
                </div>
                <div class="card-body"><div class="chart-container"><canvas id="rec-price-chart"></canvas></div></div>
            </div>
            <div class="card shadow-sm">
                <div class="card-header bg-light"><strong>Day-by-Day</strong> <small class="text-muted">— click <i class="bi bi-eye"></i> to see full context</small></div>
                <div class="card-body p-0">
                    <table class="table table-sm table-hover mb-0">
                        <thead class="table-light">
                            <tr><th>Date</th><th>Price</th><th>Ref Rate</th><th>Arm</th><th>Offset</th><th>Confidence</th><th>Actions</th></tr>
                        </thead>
                        <tbody>${this.results.map((r, i) => this.dayRow(r, i)).join('')}</tbody>
                    </table>
                </div>
            </div>`;

        document.getElementById('rec-approve-all').addEventListener('click', () => this.approveAll());
        el.querySelectorAll('.rec-ctx-btn').forEach(btn => {
            btn.addEventListener('click', () => this.showContext(parseInt(btn.dataset.idx)));
        });
        el.querySelectorAll('.rec-approve-day').forEach(btn => {
            btn.addEventListener('click', () => this.approveDay(parseInt(btn.dataset.idx)));
        });
        el.querySelectorAll('.rec-modify-day').forEach(btn => {
            btn.addEventListener('click', () => this.modifyDay(parseInt(btn.dataset.idx)));
        });
        this.renderChart();
    },

    dayRow(r, idx) {
        const d = new Date(r.stay_date);
        const day = d.toLocaleDateString('en-US', { weekday: 'short' });
        const date = d.toLocaleDateString('en-US', { month: 'short', day: 'numeric' });
        return `<tr id="rec-row-${idx}">
            <td><strong>${day}</strong> ${date}</td>
            <td><strong>${Utils.currency(r.published_price)}</strong></td>
            <td class="text-muted">${Utils.currency(r.reference_rate)}</td>
            <td>${Utils.armLabel(r.chosen_arm_label, r.chosen_arm_offset_pct)}</td>
            <td>${(r.chosen_arm_offset_pct >= 0 ? '+' : '')}${(r.chosen_arm_offset_pct * 100).toFixed(1)}%</td>
            <td>${Utils.confidenceBadge(r.confidence_score, r.confidence_label)}</td>
            <td>
                <div class="btn-group btn-group-sm">
                    <button class="btn btn-outline-info rec-ctx-btn" data-idx="${idx}" title="View full context">
                        <i class="bi bi-eye"></i>
                    </button>
                    <button class="btn btn-outline-success rec-approve-day" data-idx="${idx}" title="Approve & publish">
                        <i class="bi bi-check-lg"></i>
                    </button>
                    <button class="btn btn-outline-warning rec-modify-day" data-idx="${idx}" title="Modify price">
                        <i class="bi bi-pencil"></i>
                    </button>
                </div>
            </td>
        </tr>`;
    },

    showContext(idx) {
        const r = this.results[idx];
        const ctx = r.context || {};
        const d = new Date(r.stay_date);
        const body = document.getElementById('rec-context-modal-body');
        body.innerHTML = `
            <div class="row">
                <div class="col-md-5">
                    <h6 class="text-primary">Decision Summary</h6>
                    <table class="table table-sm">
                        <tr><td class="fw-bold">Date</td><td>${d.toLocaleDateString('en-US', {weekday:'long', month:'long', day:'numeric', year:'numeric'})}</td></tr>
                        <tr><td class="fw-bold">Price</td><td><strong>${Utils.currency(r.published_price)}</strong></td></tr>
                        <tr><td class="fw-bold">Reference Rate</td><td>${Utils.currency(r.reference_rate)}</td></tr>
                        <tr><td class="fw-bold">Arm</td><td>${Utils.armLabel(r.chosen_arm_label, r.chosen_arm_offset_pct)} (${(r.chosen_arm_offset_pct*100).toFixed(1)}%)</td></tr>
                        <tr><td class="fw-bold">Confidence</td><td>${Utils.confidenceBadge(r.confidence_score, r.confidence_label)}</td></tr>
                        <tr><td class="fw-bold">Requires Approval</td><td>${r.requires_approval ? 'Yes' : 'No'}</td></tr>
                    </table>
                    <h6 class="text-primary mt-3">Excluded Arms</h6>
                    ${r.excluded_arms && r.excluded_arms.length ? `<ul class="list-unstyled small">${r.excluded_arms.map(e => `<li><i class="bi bi-shield-x text-warning me-1"></i><strong>${e.arm.label}</strong>: ${e.reason}</li>`).join('')}</ul>` : '<p class="text-muted small">None excluded</p>'}
                    <h6 class="text-primary mt-3">All Arm Probabilities</h6>
                    <table class="table table-sm">
                        <thead><tr><th>Arm</th><th>Prob</th></tr></thead>
                        <tbody>${(r.all_arms||[]).map(a => `<tr><td>${Utils.armLabel(a.label, a.offset_pct)}</td><td>${(a.probability*100).toFixed(1)}%</td></tr>`).join('')}</tbody>
                    </table>
                </div>
                <div class="col-md-7">
                    <h6 class="text-primary">Full Context Features</h6>
                    <div class="context-panel" style="max-height:500px; overflow-y:auto;">
                        ${this.renderContextGroups(ctx)}
                    </div>
                </div>
            </div>`;
        new bootstrap.Modal(document.getElementById('rec-context-modal')).show();
    },

    renderContextGroups(ctx) {
        const groups = {
            'Demand / Operational': [
                { key: 'occupancy_pct', label: 'Occupancy %' },
                { key: 'adr_trend_pct', label: 'ADR Trend %' },
                { key: 'pace_vs_stly_pct', label: 'Pace vs STLY %' },
                { key: 'pickup_last_7d', label: 'Pickup (last 7d)' },
                { key: 'remaining_inventory_pct', label: 'Remaining Inv. %' },
            ],
            'Competitive Set': [
                { key: 'comp_set_avg_rate', label: 'Comp Set Avg Rate' },
                { key: 'our_rate_vs_compset_index', label: 'Our Rate / Compset' },
                { key: 'compset_rate_trend_pct', label: 'Compset Trend %' },
                { key: 'compset_rank', label: 'Compset Rank' },
                { key: 'compset_dispersion', label: 'Compset Dispersion' },
            ],
            'Events': [
                { key: 'event_flag', label: 'Event Active' },
                { key: 'event_intensity', label: 'Event Intensity' },
            ],
            'Segment / Calendar': [
                { key: 'segment', label: 'Segment' },
                { key: 'day_of_week', label: 'Day of Week' },
                { key: 'lead_time_days', label: 'Lead Time (days)' },
                { key: 'los_bucket', label: 'LOS Bucket' },
                { key: 'room_type', label: 'Room Type' },
                { key: 'rate_plan', label: 'Rate Plan' },
            ],
            'Identity': [
                { key: 'property_id', label: 'Property' },
                { key: 'cluster_id', label: 'Cluster' },
                { key: 'tenant_id', label: 'Tenant' },
            ],
        };
        let html = '';
        for (const [group, fields] of Object.entries(groups)) {
            html += `<div class="fw-bold text-primary mt-2 mb-1 small">${group}</div>`;
            for (const f of fields) {
                const val = ctx[f.key];
                if (val === undefined) continue;
                let display = val;
                if (typeof val === 'number') display = val.toFixed(val % 1 === 0 ? 0 : 2);
                if (typeof val === 'boolean') display = val ? '<i class="bi bi-check-circle-fill text-success"></i> Yes' : '<i class="bi bi-x-circle text-muted"></i> No';
                html += `<div class="d-flex justify-content-between py-1 border-bottom">
                    <span class="context-key small">${f.label}</span>
                    <span class="context-value small fw-bold">${display}</span>
                </div>`;
            }
        }
        return html;
    },

    renderChart() {
        if (this.chart) this.chart.destroy();
        const ctx = document.getElementById('rec-price-chart').getContext('2d');
        const labels = this.results.map(r => {
            const d = new Date(r.stay_date);
            return d.toLocaleDateString('en-US', { weekday: 'short', month: 'short', day: 'numeric' });
        });
        this.chart = new Chart(ctx, {
            type: 'line',
            data: {
                labels,
                datasets: [
                    { label: 'Recommended Price', data: this.results.map(r => r.published_price), borderColor: '#0d6efd', backgroundColor: 'rgba(13,110,253,0.1)', fill: true, tension: 0.3, pointRadius: 5 },
                    { label: 'Reference Rate', data: this.results.map(r => r.reference_rate), borderColor: '#6c757d', borderDash: [5,5], pointRadius: 2, fill: false },
                ],
            },
            options: { responsive: true, maintainAspectRatio: false, plugins: { legend: { position: 'top' } }, scales: { y: { title: { display: true, text: 'Price ($)' } } } },
        });
    },

    async approveDay(idx) {
        const r = this.results[idx];
        try {
            const payload = {
                property_id: r.property_id, room_type: r.room_type, rate_plan: r.rate_plan,
                stay_date: r.stay_date, los_nights: parseInt(document.getElementById('rec-los').value),
                dry_run: false,
            };
            // Apply same overrides so the persisted decision matches what was previewed
            const overrides = this.buildOverrides();
            if (overrides) payload.context_overrides = overrides;

            const result = await API.score(payload);
            Utils.showToast(`${Utils.formatDate(r.stay_date)}: ${Utils.currency(result.published_price)} published!`, 'success');
            const row = document.getElementById(`rec-row-${idx}`);
            if (row) { row.classList.add('table-success'); row.querySelector('.rec-approve-day').disabled = true; }
        } catch (e) { Utils.showToast('Failed: ' + e.message, 'error'); }
    },

    async modifyDay(idx) {
        const r = this.results[idx];
        const newPrice = prompt(`Modify price for ${Utils.formatDate(r.stay_date)}\nRecommended: ${Utils.currency(r.published_price)}\n\nEnter new price:`, r.published_price.toFixed(2));
        if (!newPrice || isNaN(parseFloat(newPrice))) return;
        try {
            const payload = {
                property_id: r.property_id, room_type: r.room_type, rate_plan: r.rate_plan,
                stay_date: r.stay_date, los_nights: parseInt(document.getElementById('rec-los').value),
                dry_run: false,
            };
            const overrides = this.buildOverrides();
            if (overrides) payload.context_overrides = overrides;

            const result = await API.score(payload);
            if (result.decision_id) {
                await API.override(result.decision_id, 'revenue_manager', parseFloat(newPrice));
                Utils.showToast(`${Utils.formatDate(r.stay_date)}: Overridden to ${Utils.currency(newPrice)} and published!`, 'success');
            }
            const row = document.getElementById(`rec-row-${idx}`);
            if (row) row.classList.add('table-warning');
        } catch (e) { Utils.showToast('Failed: ' + e.message, 'error'); }
    },

    async approveAll() {
        if (!confirm(`Approve and publish all ${this.results.length} days?`)) return;
        let success = 0;
        for (let i = 0; i < this.results.length; i++) {
            try { await this.approveDay(i); success++; } catch(e) {}
        }
        Utils.showToast(`Published ${success}/${this.results.length} recommendations.`, 'success');
    },

    avg(field) {
        if (!this.results.length) return 0;
        return this.results.reduce((s, r) => s + r[field], 0) / this.results.length;
    },
};
